keycloak-event-emitter/src/main/java/io/cloudtrust/keycloak/kafkaeventemitter/KafkaEventEmitterProvider.java

```diff
@@ -0,0 +1,129 @@
+import io.cloudtrust.keycloak.customevent.IdentifiedEvent;
+import io.cloudtrust.keycloak.snowflake.IdGenerator;
+import org.apache.kafka.clients.producer.Producer;
+import org.apache.kafka.clients.producer.ProducerRecord;
+import org.jboss.logging.Logger;
+import org.keycloak.events.Details;
+import org.keycloak.events.Event;
+import org.keycloak.events.EventListenerProvider;
+import org.keycloak.events.admin.AdminEvent;
+import org.keycloak.models.KeycloakSession;
+import org.keycloak.models.RealmModel;
+import org.keycloak.models.UserModel;
+
+import java.nio.ByteBuffer;
+import java.util.Base64;
+import java.util.HashMap;
+import java.util.regex.Matcher;
+import java.util.regex.Pattern;
+
+public class KafkaEventEmitterProvider implements EventListenerProvider {
+    private static final Logger logger = Logger.getLogger(KafkaEventEmitterProvider.class);
+
+    private final Producer<String, String> producer;
+    private final KeycloakSession keycloakSession;
+    private final String eventTopic;
+    private final String adminEventTopic;
+    private final IdGenerator idGenerator;
+
+    KafkaEventEmitterProvider(KeycloakSession keycloakSession, Producer<String, String> producer, String eventTopic, String adminEventTopic, IdGenerator idGenerator) {
+        this.keycloakSession = keycloakSession;
+        this.producer = producer;
+        this.eventTopic = eventTopic;
+        this.adminEventTopic = adminEventTopic;
+        this.idGenerator = idGenerator;
+    }
+
+    @Override
+    public void onEvent(Event event) {
+        completeEventAttributes(event);
+        long uid = idGenerator.nextValidId();
+        IdentifiedEvent identifiedEvent = new IdentifiedEvent(uid, event);
+
+        // Flatbuffer serialization
+        ByteBuffer buffer = SerialisationUtils.toFlat(identifiedEvent);
+
+        produceEvent(buffer, eventTopic);
+    }
+
+    @Override
+    public void onEvent(AdminEvent adminEvent, boolean includeRepresentation) {
+        long uid = idGenerator.nextValidId();
+        IdentifiedAdminEvent identifiedAdminEvent = new IdentifiedAdminEvent(uid, adminEvent);
+        ExtendedAdminEvent customAdminEvent = completeAdminEventAttributes(identifiedAdminEvent);
+
+        // Flatbuffer serialization
+        ByteBuffer buffer = SerialisationUtils.toFlat(customAdminEvent);
+
+        produceEvent(buffer, adminEventTopic);
+    }
+
+    @Override
+    public void close() {
+
+    }
+
+    private void completeEventAttributes(Event event) {
+        // add username if missing
+        if (event.getDetails() == null) {
+            event.setDetails(new HashMap<>());
+        }
+        String eventUsername = event.getDetails().get(Details.USERNAME);
+        if (!Strings.isNullOrEmpty(event.getUserId()) && Strings.isNullOrEmpty(eventUsername)) {
+            RealmModel realm = keycloakSession.realms().getRealm(event.getRealmId());
+            // retrieve username from userId
```

Comment:
realm!=nil ? (even if I guess this is a copy of the legacy EventEmitter)

Reply:
Not sure it is a case that can happen, but I added this check

Reply:
I'm not sure it can happen too... it's just in case we are really not lucky...